theano/tensor/basic.py

```diff
@@ -4822,33 +4822,45 @@ def infer_shape(self, node, ishapes):
             return [(1,) * self.ndim]
 
         requ = node.inputs[1]
+        input_size = mul(*ishapes[0])
         if isinstance(requ, theano.tensor.TensorConstant):
             requ = list(requ.data)
             requ_part = [ele for ele in requ if ele != -1]
             crit = len(requ) - len(requ_part)
             if crit == 1 and len(requ_part) > 0:
-                missing = mul(*ishapes[0]) // mul(*requ_part)
+                # If there are both 0 and -1 in requ_size, it is impossible
+                # to determine a right output, but we can at least prevent
+                # a division by 0.
+                requ_size = mul(*requ_part)
+                missing = input_size // (1 if requ_size == 0 else requ_size)
```

Comment:
Right, if we know that in advance, we can do that instead of mimicking the symbolic case.
However, a shape of -1 is not valid, and it could lead to weird failures, like alloc(..., -1), or min(-1, 0) getting replaced by 0.